src/schematools/permissions/auth.py

```diff
@@ -0,0 +1,334 @@
+    """
+
+    def __init__(self):
+        self._collection = []
+
+    def append(self, permission: Permission):
+        """Add a new discovered permission level to this collection."""
+        self._collection.append(permission)
+        if permission.level is PermissionLevel.highest:
+            # Avoid further code flow.
+            raise HighestPermissionFound(permission)
+
+    @property
+    def highest_value(self) -> Permission:
+        """Return the highest found permission."""
+        return max(self._collection, default=Permission.none)
+
+    def __bool__(self):
+        return bool(self._collection)
+
+
+class UserScopes:
+    """A request-like object that tells what the current user may access.
+    This is the foundation for all permission checks.
+
+    All ``has_...()`` functions are used for permission checks.
+    Internally, these read the schema and profile data for the authorization matrix.
+
+    * By default, all fields can be read unless the schema defines an "auth" field.
+    * The "auth" flags in schema files act as a blacklist: no access, except for some roles.
+    * The "profile" rules open up certain fields, hence whitelist features.
+    """
+
+    def __init__(
+        self,
+        query_params: dict[str, ...],
+        is_authorized_for: HasAllScopesFunc,
+        all_profiles: Optional[Iterable[ProfileSchema]] = None,
+    ):
+        """
+        :param query_params: The search query filter (e.g. request.GET).
+        :param is_authorized_for: A function that checks for authorization of scopes
+            For example ``request.is_authorized_for()`` from ``authorization_django``
+            or use :func:`create_scopes_check` as a substitute.
+        :param all_profiles: All profiles that need to be loaded.
+        """
+        self._query_param_names = [param for param, value in query_params.items() if value]
+        self._all_profiles = all_profiles
+
+        # Reuse existing function (e.g. request.is_authorized_for() by authorization_django)
+        # The has_..._scopes functions are both cached within this request for performance.
+        self.has_all_scopes = lru_cache()(is_authorized_for)
+
+    def load_profiles(self):
+        """Import a list of profiles, this only stores the profiles that match the scope."""
+        self._all_profiles = [p.schema for p in Profile.objects.all()]
+
+    def add_query_params(self, params: List[str]):
+        """Tell that the request has extra (implicit) parameters that are satisfied, e.g. via the URL."""
+        self._query_param_names.extend(params)
+
+    @methodtools.lru_cache()
+    def has_all_scopes(self, *needed_scopes: str) -> bool:
+        """Check whether the request has all scopes.
+        Note: at ``__init__()``, this function receives an LRU-cache for performance.
+        """
+        # This stub exists for documentation purposes only. It's replaced by the init parameter.
+        raise RuntimeError("this method will be overwritten")
```

Comment:
Please use ` NotImplementedError` here, as not implementing this method means developer error, not process error.

Reply:
Good point. It was a `RuntimeError` as the method should be overwritten in `__init__` but I've removed that magic instead.